python/tk_multi_publish2/api/manager.py

```diff
@@ -0,0 +1,657 @@
+
+    # this is the key we'll use to store a special property on items that
+    # are collected via a file path. we can use this later on to determine which
+    # items were added to the tree via path collection and what that original
+    # path was (client code could add multiple items for a single path).
+    PROPERTY_KEY_COLLECTED_FILE_PATH = "__collected_file_path__"
+
+    ############################################################################
+    # instance methods
+
+    def __init__(self, publish_logger=None):
+        """
+        Initialize the manager.
+
+        :param publish_logger: This is a standard python logger to use during
+            publishing. A default logger will be provided if not supplied. This
+            can be useful when implementing a custom UI, for example, with a
+            specialized log handler (as is the case with the Publisher)
+        """
+
+        # the current bundle (the publisher instance)
+        self._bundle = sgtk.platform.current_bundle()
+
+        # a logger to be used by the various collector/publish plugins
+        self._logger = publish_logger or logger
+
+        # the underlying tree representation of the items to publish
+        self._tree = PublishTree()
+
+        # collector instance for this context
+        self._collector_instance = None
+
+        # a lookup of context to publish plugins.
+        self._processed_contexts = {}
+
+        # initialize the collector plugin
+        logger.debug("Loading collector plugin...")
+        self._load_collector()
+
+        # go ahead and load the plugins for the current context for efficiency
+        logger.debug("Loading plugins for the current context...")
+        self._load_publish_plugins(self._bundle.context)
+
+        # load the phose phase hook
+        logger.debug("Loading post phase hook...")
+        post_phase_hook_path = self._bundle.get_setting(
+            self.CONFIG_POST_PHASE_HOOK_PATH)
+        self._post_phase_hook = self._bundle.create_hook_instance(
+            post_phase_hook_path,
+            base_class=self._bundle.base_hooks.PostPhaseHook
+        )
+
+    def collect_files(self, file_paths):
+        """
+        Run the collection logic to populate the publish tree with items for
+        each supplied path.
+
+        Each path supplied to this method will be processed by the configured
+        collector hook for the current context. The collector will create
+        :ref:`publish-api-item` instances accordingly, each of which will be
+        marked as :py:attr:`~.api.PublishItem.persistent`.
+
+        :param list file_paths: A list of file paths to collect as items to
+            publish.
+        :returns: A list of the created :ref:`publish-api-item` instances.
+        """
+
+        new_items = []
+
+        for file_path in file_paths:
+
+            # get a list of all items in the tree prior to collection
+            items_before = list(self.tree.items)
```

Comment:
We went to the trouble of making these generators, but then we just flatten them out into lists as soon as we need them? Should we just ditch the generators and return lists to remove some logic?

Reply:
I think the intent with the generators was to not assume that the tree will always have to be iterated completely over.

So I'm fine with something returning an iterator. What I'm increasingly getting less comfortable with is the name of the methods that return them, especially the __iter__ on the item. That one is not obvious to me. Iterating on a tree should iterate over all nodes, but iterating on a node should probably visit it's direct children, like the `item.children` property.